Added support for previous and next containing scope type

AndreasArvidsson opened the pull request:
`"take next funk"`
`"change previous state"`

For now you can't use this when not in the current containing scope. This is a limitation of our current implication of every. I still feel it could be really useful.

Fixes https://github.com/cursorless-dev/cursorless/issues/982

## Checklist

- [x] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)

AndreasArvidsson commented on cursorless-talon/src/modifiers/containing_scope.py:
```diff
@@ -84,16 +84,24 @@ def cursorless_custom_regex_scope_type(m) -> dict[str, str]:
 
 @mod.capture(
-    rule="[every] (<user.cursorless_scope_type> | <user.cursorless_custom_regex_scope_type>)"
+    rule="[every | previous | next] (<user.cursorless_scope_type> | <user.cursorless_custom_regex_scope_type>)"
```
We really should move these into a csv file. Any recommendation for which file?

auscompgeek commented on src/typings/targetDescriptor.types.ts:
```diff
@@ -171,6 +171,16 @@ export interface ContainingScopeModifier {
   scopeType: ScopeType;
 }
 
+export interface PreviousContainingScopeModifier {
+  type: "previousContainingScope";
+  scopeType: ScopeType;
+}
+
+export interface NextContainingScopeModifier {
+  type: "nextContainingScope";
+  scopeType: ScopeType;
+}
```
Is having these as separate interfaces useful? Maybe we can simplify this to be a single interface. Something like this

```ts
export interface SiblingContainingScopeModifier {
  type: "previousContainingScope" | "nextContainingScope";
  scopeType: ScopeType;
}
```

AndreasArvidsson commented:
@phillco @auscompgeek I have now updated previous/next to use existing ordinal range modifier. For now I have removed all changes Talon side and is going to do a follow up pr regarding #13 where I'm going to implement all the grammar and thoroughly test these.

I think we need discussion about how to handle relative sub tokens with regards to explicit range. For now relative is not enabled for sub tokens.

pokey commented on src/processTargets/modifiers/scopeTypeStages/SubTokenStages.ts:
```diff
@@ -0,0 +1,139 @@
+
+  private findSingleTarget(target: Target, targets: Target[]): Target {
+    let intersectingTargets = targets
+      .map((t) => ({
+        target: t,
+        intersection: t.contentRange.intersection(target.contentRange),
+      }))
+      .filter((it) => it.intersection != null);
```
why do you need the `map` here?  why not just filter directly
```suggestion
      .filter((t) => t.contentRange.intersection(target.contentRange) != null);
```

AndreasArvidsson replied:
Because I require the intersection at a later stage

pokey commented on src/processTargets/modifiers/scopeTypeStages/SubTokenStages.ts:
```diff
@@ -0,0 +1,139 @@
+    return [this.findSingleTarget(target, targets)];
+  }
+
+  private findSingleTarget(target: Target, targets: Target[]): Target {
```
I'd rename `targets` to something like `allTargets` or something.  Maybe look at your `ItemStage` code for naming inspiration

pokey commented on src/processTargets/modifiers/scopeTypeStages/SubTokenStages.ts:
```diff
@@ -0,0 +1,139 @@
+    }
+
+    // On non empty range utilize all non-empty intersecting targets
+    intersectingTargets = intersectingTargets.filter(
+      (it) => !it.intersection!.isEmpty
+    );
```
is this what we do for eg "token", "block" and "line"?  I'd be tempted to expand on adjacent but idk.  Maybe worth discussing at meet-up

I see now why you're doing the `map` above.  Might be able to remove it depending what we decide here

AndreasArvidsson replied:
That would mean that if you have a single character selected we would expand to all three of them. Probably not to the desired behavior

pokey commented on src/processTargets/modifiers/scopeTypeStages/SubTokenStages.ts:
```diff
@@ -0,0 +1,139 @@
+    super(modifier, SUBWORD_MATCHER);
+  }
+
+  protected createTargets(target: Target, ranges: Range[]): Target[] {
```
I would call this `inputTarget`.  We're throwing the word "target" around a lot here so a bit of verbosity might aid clarity

I'd also add a docstring; not immediately obvious what this does and what `ranges` is supposed to be

pokey commented on src/processTargets/processTargets.ts:
```diff
@@ -246,3 +235,9 @@ function calcIsReversed(anchor: Target, active: Target) {
+
+function ensureSingleEditor(anchorTarget: Target, activeTarget: Target) {
+  if (anchorTarget.editor !== activeTarget.editor) {
+    throw new Error("AnchorTarget and activeTarget must be in same document");
```
```suggestion
    throw new Error("Cannot form range between targets in different editors");
```

pokey commented on src/test/suite/fixtures/recorded/selectionTypes/clearCar.yml:
```diff
@@ -0,0 +1,23 @@
+      active: {line: 0, character: 4}
+  marks: {}
+finalState:
+  documentContents: aaabbCcc
```
huh.  I think I'd go right here.  That's what we do for surrounding pair

AndreasArvidsson replied:
I follow the logic we use with token. There we prefer left.

pokey reviewed (commented):
Wow this turned out super clean!  Left a million little comments and a couple things to discuss at meet-up 😅

AndreasArvidsson commented on cursorless-talon/src/modifiers/ordinals.py:
```diff
@@ -0,0 +1,75 @@
+            "excludeAnchor": not include_anchor,
+            "excludeActive": not include_active,
+        }
+    except AttributeError:
```
Agreed

AndreasArvidsson commented on src/core/commandVersionUpgrades/upgradeV1ToV2/upgradeV1ToV2.ts:
```diff
@@ -155,7 +156,7 @@ function upgradePrimitiveTarget(
     mark: mark?.type === "cursorToken" ? undefined : mark,
     // Empty array of modifiers is not allowed
-    modifiers: modifiers.length > 0 ? modifiers : undefined,
+    modifiers: modifiers.length > 0 ? (modifiers as Modifier[]) : undefined,
```
Because they are of different types. We had to do a new version of modifier when we retired ordinal range.

AndreasArvidsson commented on src/typings/targetDescriptor.types.ts:
```diff
@@ -248,19 +266,29 @@ export interface CascadingModifier {
   modifiers: Modifier[];
 }
 
+export interface RangeModifier {
+  type: "range";
+  anchor: Modifier;
+  active: Modifier;
+  excludeAnchor?: boolean;
+  excludeActive?: boolean;
+}
```
Sounds good. I will be there.

AndreasArvidsson commented on src/core/commandVersionUpgrades/upgradeV2ToV3/upgradeV2ToV3.ts:
```diff
@@ -0,0 +1,100 @@
+  target: PartialPrimitiveTargetDescriptor
+): PartialPrimitiveTargetDescriptor {
+  if (target.modifiers != null) {
+    const modifiers = target.modifiers as ModifierV2[];
```
The old modifier is only available in version two and the new one only in version three

AndreasArvidsson commented on src/core/commandVersionUpgrades/upgradeV2ToV3/upgradeV2ToV3.ts:
```diff
@@ -0,0 +1,100 @@
+    for (let i = 0; i < modifiers.length; ++i) {
+      const oldModifier = modifiers[i];
+      if (oldModifier.type === "ordinalRange") {
+        target.modifiers[i] = createNewModifier(oldModifier);
```
I can use a map() if you prefer? I don't really see a problem with the current approach thou

AndreasArvidsson commented on src/processTargets/modifiers/scopeTypeStages/SubTokenStages.ts:
```diff
@@ -0,0 +1,139 @@
+import { Range } from "vscode";
```
Because sub words has special logic around delimiters that the regexStage is lacking

AndreasArvidsson commented on src/processTargets/modifiers/OrdinalStages.ts:
```diff
@@ -0,0 +1,118 @@
+  run(context: ProcessedTargetsContext, target: Target): Target[] {
+    const targets = getTargets(
+      context,
+      createTargetWithoutExplicitRange(target),
```
I feel the same

AndreasArvidsson commented on src/test/suite/fixtures/recorded/selectionTypes/clearWord.yml:
```diff
@@ -0,0 +1,23 @@
+languageId: plaintext
+command:
```
Could you elaborate on the first test?

AndreasArvidsson commented on src/processTargets/modifiers/OrdinalStages.ts:
```diff
@@ -0,0 +1,112 @@
+        index: i,
+        intersection: t.contentRange.intersection(target.contentRange),
+      }))
+      .filter((t) => t.intersection != null)
```
In case of an empty selection we would get no containing indices. ie empty selection in the middle of a function will give no result which is not what we want.

AndreasArvidsson commented on src/processTargets/modifiers/OrdinalStages.ts:
```diff
@@ -0,0 +1,112 @@
+import { Target } from "../../typings/target.types";
```
It is less than 150 lines so I think it's fine in a single file? Deep directory hierarchies can also be annoying.

pokey reviewed (commented):
- [ ] make sure that "token" goes to the right when the tokens are equivalent

pokey commented on src/test/suite/fixtures/recorded/selectionTypes/takeEveryCarBlock.yml:
```diff
@@ -0,0 +1,51 @@
+      active: {line: 0, character: 2}
+    - anchor: {line: 0, character: 2}
+      active: {line: 0, character: 3}
+    - anchor: {line: 0, character: 3}
+      active: {line: 1, character: 0}
```
For some reason "char" seems to be yielding a newline character here

AndreasArvidsson replied:
The problem is that `/\p{L}\p{M}*|\P{L}/gu` matches whitespace and new lines as well

pokey replied:
That is so strange. Why on earth would whitespace be considered a letter or a diacritic?

AndreasArvidsson replied:
Sounds reasonable. Can we get by just with? `/\p{L}\p{M}*/` or do you have any better suggestions?

pokey replied:
I think that will miss symbols like brackets. I'd look at the token regex and see what we can steal from there. If you come up blank I can take a look a bit later

AndreasArvidsson replied:
This seems to work both grapheme and sub token `/\p{L}\p{M}*|[^\p{L}\p{Z}\p{C}]/gu`

pokey reviewed (commented):
- [ ] Add migration and talon side for getting "row" to use new representation; don't bother with ranges for now

AndreasArvidsson commented on src/test/suite/fixtures/recorded/subtoken/takeFirstChar5.yml:
```diff
@@ -16,6 +16,6 @@ initialState:
   selections:
-    - anchor: {line: 0, character: 0}
-      active: {line: 0, character: 1}
+    - anchor: {line: 0, character: 2}
```
@pokey Notice that we now select the token to the right that will change current behavior

AndreasArvidsson replied:
The problem is that `/\p{L}\p{M}*|\P{L}/gu` matches whitespace and new lines as well

AndreasArvidsson commented:
@pokey Ready for another review! :)

pokey reviewed (commented):
- [ ] Change `type` fields to `"relativeScope"` and `"ordinalScope"`

AndreasArvidsson commented:
@pokey Modifiers are renamed

pokey commented on src/processTargets/modifiers/ordinalStages/RelativeScopeStage.ts:
```diff
@@ -0,0 +1,99 @@
+
+  // Content range is empty. Use rightmost target and accept weak containment.
+  if (inputTargetRange.isEmpty) {
+    // FIXME: Handle case where no targets intersect with the input range
```
This breaks when the input is between targets.  For example:

```
hello | world
```

In this case `"next token"` breaks

AndreasArvidsson replied:
I know. My reasoning for leaving it as it is was that most of the scopes wouldn't even support `every` in that case and therefore relative scope has no targets to work with.

pokey replied:
Right, but "next token" should work, and is pretty useful

Also I think it indicates a flaw in the logic of the code as it violates an assumption that is baked in. Hesitant to merge the code with this issue

AndreasArvidsson replied:
Okay.  I will have a look at it tomorrow.

AndreasArvidsson replied:
done

pokey reviewed (changes requested):
Ok I pushed a bunch of minor changes, and was hoping to just do everything myself, but ran into an issue with the logic for relative scopes, and I think it may require a rethinking of the way you handle relative scopes, so I'm going to put this one back into your court

pokey reviewed (commented):
Ok I made a bunch of minor changes, and left one question.  Lmk if you have a problem with any of the changes, and curious to hear your thoughts on my comment, but otherwise good to go!

AndreasArvidsson commented on src/processTargets/modifiers/ordinalStages/RelativeScopeStage.ts:
```diff
@@ -0,0 +1,138 @@
+
+      // The content range isnt containing any scopes, but they should not be
+      // included anyhow. Find a non containing relative index.
+      if (containingIndices.length === 0 && !includeContaining) {
```
I'm not against is. Looks like you are on a roll so please go ahead and make these changes :)

pokey replied:
done :)

pokey reviewed (approved):
Ok I think this is good to go!  I added a bunch of tests and did some cleanup.  Have a quick look before you merge to make sure you're happy

Fwiw I made a super basic grammar for `"two funk"` so I could easily add tests; doesn't have pluralisation tho so I'm not going to merge it.  [Here](https://github.com/cursorless-dev/cursorless/commit/6820a2104cd03a62a019dd970405498c5c286f4d) it is tho in case helpful

AndreasArvidsson commented:
I think it looks good :)
